contracts/staking/Staking.sol

```diff
@@ -129,212 +128,165 @@ contract Staking is
-        BundleStakeInfo storage stakeInfo = _bundleStakeInfo[instanceId][bundleId][user];
-
-        uint256 rewards = calculateRewardsIncrement(stakeInfo);
-        if(rewards > 0) {
-            _increaseBundleStakes(stakeInfo, rewards);
-        }
+        Target memory target = _target[targetId];
+        StakeInfo storage info = _stakeInfo[targetId][user];
 
-        bool unstakeAll = (amount == type(uint256).max);
-        if(unstakeAll) {
-            amount = stakeInfo.balance;
+        // handling for new stakes
+        if(info.createdAt == 0) {
+            info.user = user;
+            info.targetId = targetId;
+            info.stakeBalance = 0;
+            info.rewardBalance = 0;
+            info.createdAt = block.timestamp;
         }
 
-        _decreaseBundleStakes(stakeInfo, amount);
-        _payoutStakes(user, amount);
+        _updateRewards(target, info);
+        _increaseStakes(info, amount);
+        _collectDip(user, amount);
 
-        emit LogStakingUnstakedFromBundle(
-            user,
-            instanceId,
-            bundleId,
+        emit LogStakingStaked(
+            info.user,
+            info.targetId,
+            target.instanceId,
+            target.componentId,
+            target.bundleId,
             amount,
-            rewards,
-            unstakeAll
+            _stakeBalance
         );
     }
 
 
-    function setRewardRate(
-        uint256 newRewardRate
-    )
-        external override
-        onlyOwner
-    {
-        require(newRewardRate <= _rewardRateMax, "ERROR:STK-060:REWARD_EXCEEDS_MAX_VALUE");
-        uint256 oldRewardRate = _rewardRate;
+    function unstake(bytes32 targetId, uint256 amount) external override {
+        _unstake(targetId, msg.sender, amount);
+    }
 
-        _rewardRate = newRewardRate;
 
-        emit LogStakingRewardRateSet(oldRewardRate, newRewardRate);
+    function unstakeAndClaimRewards(bytes32 targetId) external override {
+        _unstake(targetId, msg.sender, type(uint256).max);
     }
 
 
-    function hasBundleStakeInfo(
-        bytes32 instanceId,
-        uint256 bundleId,
-        address user
-    )
+    function claimRewards(bytes32 targetId)
         external override
-        view
-        returns(bool hasInfo)
+        onlyInfo(targetId, msg.sender)
     {
-        BundleStakeInfo memory stakeInfo = _bundleStakeInfo[instanceId][bundleId][user];
-        return stakeInfo.createdAt > 0;
+        Target memory target = _target[targetId];
+        StakeInfo storage info = _stakeInfo[targetId][msg.sender];
+        _claimRewards(target, info);
     }
 
 
-    function getBundleStakeInfo(
-        bytes32 instanceId, 
-        uint256 bundleId,
-        address user
-    )
-        external override
-        view
-        onlyWithBundleStakeInfo(instanceId, bundleId, user)
-        returns(BundleStakeInfo memory info)
+    // only for testing purposes!
+    // decide if this should be restricted to ganache chain ids
```

Comment:
i think thats a good idea. ganache and mumbai. 
and maybe we should document somewhere all places we set this restriction. This would help us in case we ever deploy to another test network instead of mumbai.

Reply:
maybe we should consider registring chains with an attribute if it's a prod or test chain, from there this function could require that the call is only allowed for test chains

Reply:
i like that approach.